icu4c/source/i18n/collationdatabuilder.cpp

```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
     setDigitTags(errorCode);
     setLeadSurrogates(errorCode);
 
-    if (!icu4xMode) {
+    if (icu4xMode) {
+        // Make sure that starters that occur is the middle of a
+        // contraction have contraction ce32 with the
+        // `CONTRACT_HAS_STARTER` flag set so that starters that
+        // can occur in a non-final position in a contraction can
+        // be easily recognized from having a contraction ce32
+        // that has the `CONTRACT_HAS_STARTER` flag set.
+
+        UCharsTrieBuilder contractionBuilder(errorCode);
+        // Intentionally unpaired low surrogate to make it never
+        // match well-formed UTF-16 which ICU4X feeds to the
+        // matcher.
+        UnicodeString placeholder(0xDC00);
+
+        UnicodeSetIterator iter(contractionMiddleStarter);
+        while(iter.next() && !iter.isString()) {
```

Comment:
optional but nice to have:
something like this should work since ICU 76 --
```
for (UChar32 c : contractionMiddleStarter.codePoints()) {
```

see https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#a537a13a1b3adcaebc140f715876aca63

Reply:
Changed.